wallet/src/signer/ledger_signer/ledger_messages.rs

```diff
@@ -0,0 +1,303 @@
+// Copyright (c) 2025 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{collections::BTreeMap, mem::size_of_val, time::Duration};
+
+use crate::signer::{ledger_signer::LedgerError, SignerError, SignerResult};
+use common::{
+    chain::{self},
+    primitives,
+};
+use crypto::key::{
+    extended::ExtendedPublicKey,
+    hdkd::{
+        chain_code::{ChainCode, CHAINCODE_LENGTH},
+        derivation_path::DerivationPath,
+    },
+    secp256k1::{extended_keys::Secp256k1ExtendedPublicKey, Secp256k1PublicKey},
+};
+use serialization::{Decode, DecodeAll, Encode};
+use utils::ensure;
+
+use ledger_lib::Exchange;
+use mintlayer_ledger_messages::{
+    decode_all as ledger_decode_all, encode as ledger_encode, AddrType, Amount as LAmount,
+    Bip32Path as LedgerBip32Path, CoinType, InputAdditionalInfoReq, Ins,
+    OutputValue as LOutputValue, P1SignTx, PubKeyP1, PublicKeyReq, SignMessageReq, SignTxReq,
+    TxInput as LTxInput, TxInputReq, TxMetadataReq, TxOutput as LTxOutput, TxOutputReq, APDU_CLASS,
+    H256 as LH256, P1_APP_NAME, P1_SIGN_NEXT, P1_SIGN_START, P2_DONE, P2_SIGN_MORE,
+};
+
+const MAX_ADPU_LEN: usize = (u8::MAX - 5) as usize; // 4 bytes for the header + 1 for len
+const TIMEOUT_DUR: Duration = Duration::from_secs(100);
+const OK_RESPONSE: u16 = 0x9000;
+const TX_VERSION: u8 = 1;
+
+#[derive(Decode)]
+pub struct LedgerSignature {
+    pub signature: [u8; 64],
+    pub multisig_idx: Option<u32>,
+}
+
+struct SignatureResult {
+    sig: LedgerSignature,
+    input_idx: usize,
+    has_more_signatures: bool,
+}
+
+/// Check that the response ends with the OK status code and return the rest of the response back
+pub fn ok_response(mut resp: Vec<u8>) -> SignerResult<Vec<u8>> {
+    let (_, status_code) = resp.split_last_chunk().ok_or(LedgerError::InvalidResponse)?;
+    let response_status = u16::from_be_bytes(*status_code);
+
+    ensure!(
+        response_status == OK_RESPONSE,
+        LedgerError::ErrorResponse(response_status)
```

Comment:
Let's parse the status code (e.g. using ledger-proto's `StatusCode`, which implements `TryFromPrimitive`) and provide a more informative error instead of this `ErrorResponse(u16)`. E.g. currently I'm getting "Ledger error: Received an error response from the Ledger device: 21781" when trying to connect to a locked device, this is not very informative.